internal/telemetry/region_pusher.go

```diff
@@ -0,0 +1,147 @@
+package telemetry
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	sm "github.com/grafana/synthetic-monitoring-agent/pkg/pb/synthetic_monitoring"
+	"github.com/rs/zerolog"
+)
+
+// RegionPusher periodically sends telemetry data for a specific region.
+type RegionPusher struct {
+	client sm.TelemetryClient
+	logger zerolog.Logger
+
+	instance string
+	regionID int32
+
+	telemetry   map[int64]map[sm.CheckClass]*sm.CheckClassTelemetry // Indexed by local tenant ID
+	telemetryMu sync.Mutex
+}
+
+// NewRegionPusher builds a new RegionPusher.
+// Notice that the effective time span used to dictate the pace for periodic
+// push events will be defined based on the given time span plus a random
+// jitter [0,59)s.
+func NewRegionPusher(
+	ctx context.Context, timeSpan time.Duration,
+	client sm.TelemetryClient, logger zerolog.Logger, instance string, regionID int32,
+	opts ...any,
+) *RegionPusher {
+	tp := &RegionPusher{
+		client:    client,
+		logger:    logger,
+		instance:  instance,
+		regionID:  regionID,
+		telemetry: make(map[int64]map[sm.CheckClass]*sm.CheckClassTelemetry),
+	}
+
+	var ticker ticker = newStdTicker(timeSpan)
+
+	for _, o := range opts {
+		switch o := o.(type) { //nolint:gocritic
+		case withTicker:
+			ticker = o
+		}
+	}
+
+	go tp.run(ctx, ticker)
+
+	return tp
+}
+
+type withTicker ticker
+
+func (p *RegionPusher) run(ctx context.Context, ticker ticker) {
+	p.logger.Info().Msg("region pusher starting")
+
+	// TODO: We could potentially create here a pushCtx, pass that to push, and
+	// from push keep retrying sending the data if it fails initially until the
+	// pushCtx is canceled, which should happen once the ticker ticks again, to
+	// avoid overlapping push requests.
+	// By now, only retry folllowing the ticker's pace.
+
+LOOP:
+	for {
+		select {
+		case <-ticker.C():
+			p.logger.Info().Msg("pushing telemetry")
+			m := p.next()
+			go p.push(m) // Avoid blocking
+		case <-ctx.Done():
+			p.logger.Debug().Msg("region pusher stopping")
+			m := p.next()
+			p.push(m)
+			break LOOP
+		}
+	}
+}
+
+// AddExecution adds a new execution to the tenant telemetry.
+func (p *RegionPusher) AddExecution(e Execution) {
```

Comment:
I almost want to ask for some observability here, too.

It should be enough to have a summary, to keep track of the number of calls to this function and their duration. I would prefer a (native) histogram.

The concern is that I suspect there might be a lot of contention for this lock, but it doesn't make sense to worry about that without some data. The happy path seems safe enough (O(1)).

Reply:
That makes total sense, let's add it so we can measure and adjust if required.
See [7c78455](https://github.com/grafana/synthetic-monitoring-agent/pull/621/commits/7c78455a4aef339e3cd9031d8180f807a045d635). Even though we are using native histogram, I did set also the regular buckets defining custom values that seem sensible for our needs, as we are not measuring "network requests" here, so the default buckets are probably off for this case. Feedback is welcome in that aspect. Nevertheless we can adjust once we test this at scale.

Reply:
I need to validate that the set up in the POP clusters is actually scraping native histograms, I don't think that's the case. The Grafana Agent needs to have its configuration in flow mode in order to support that.